File: Python/Leetcode/FourSum/four_sum.py
# finding three sum 
# Time complexity O(N^2)
from operator import le


def sub_three_sum(nums, target,starting_point):
    result = []
    for left in range(starting_point+1,len(nums)-2):
        # checking duplicates in first loop
        if left>starting_point+1 and nums[left]==nums[left-1]:
            continue
        mid = left + 1
        right = len(nums)-1
        while mid<right:
            sum_of_all = nums[starting_point]+nums[left]+nums[mid]+nums[right]
            if sum_of_all>target:
                right -= 1
            elif sum_of_all <target:
                mid +=1
            else:
                result.append([nums[starting_point],nums[left],nums[mid],nums[right]])

                # two while loops are for checking duplicates in a particular iteration
                while mid<right and nums[mid]==nums[mid+1]:
                    mid +=1
                while mid<right and nums[right]==nums[right-1]:
                    right -=1
                mid +=1
                right -=1
    return result

# Time complexity O(N)
def four_sum(nums, target):
    nums.sort()
    result=[]
    for i in range(len(nums)-3):
      #checking duplicates
        if i>0 and nums[i]==nums[i-1]:
            continue
        result.extend(sub_three_sum(nums,target,i))
    return result

if __name__ == "__main__":
  nums = [2,2,2,2]
  print(four_sum(nums,8))

File: Leetcode/18. 4Sum/four_sum.rs
struct Solution {}

impl Solution {
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>>{
        let mut res: Vec<Vec<i32>> = Vec::new();
        if nums.len()<4{
            return res;
        }
        let mut nums = nums;
        nums.sort();
        println!("{:?}",nums);
        for i in 0..nums.len()-3{
            if i>0 && nums[i]==nums[i-1]{
                continue;
            }
            for j in i+1..nums.len()-2{
                if j>i+1 && nums[j]==nums[j-1] {
                    continue;
                }
                let mut left = j+1;
                let mut right = nums.len()-1;
                while left<right{
                    let four_sum = nums[i] as i64+nums[j] as i64+nums[left]as i64+nums[right] as i64;
                    if four_sum==target as i64{
                        res.push(vec![nums[i],nums[j],nums[left],nums[right]]);
                        while left<right && nums[left]==nums[left+1]{
                            left+=1;
                        }
                        while left<right && nums[right]==nums[right-1]{
                            right-=1;
                        }
                        left+=1;
                        right-=1;
                    } else if four_sum>target as i64 {
                        right-=1;
                    } else{
                        left+=1;
                    }
                }

            }
        }
        res
    }
}
fn main() {
    let res = Solution::four_sum(vec![1000000000,1000000000,1000000000,1000000000],-294967296);
    println!("{:?}", res);
}
